rust/oak_log/src/lib.rs

```diff
@@ -0,0 +1,55 @@
+//
+// Copyright 2019 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! A logger that send output to an Oak logging channel.
+
+extern crate log;
+extern crate oak;
+
+use log::{Level, Log, Metadata, Record, SetLoggerError};
+use std::io::Write;
+
+struct OakChannelLogger {
+    level: Level,
+}
+
+impl Log for OakChannelLogger {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() <= self.level
+    }
+    fn log(&self, record: &Record) {
+        if !self.enabled(record.metadata()) {
+            return;
+        }
+        writeln!(
+            oak::logging_channel(),
+            "{}  {} : {} : {}",
+            record.level(),
+            record.file().unwrap_or_default(),
+            record.line().unwrap_or_default(),
+            record.args()
+        )
+        .unwrap();
+    }
+    fn flush(&self) {}
+}
+
+pub fn init(level: Level) -> Result<(), SetLoggerError> {
+    let logger = OakChannelLogger { level };
+    log::set_boxed_logger(Box::new(logger))?;
+    log::set_max_level(level.to_level_filter());
```

Comment:
Not sure why we need to keep the level in the logger but also set it globally? Maybe we should just rely on `log::max_level()` entirely instead of storing it in our own logger too? not sure

Reply:
In fact (https://docs.rs/log/0.4.6/src/log/lib.rs.html#1064):

> /// The [`log!`], [`error!`], [`warn!`], [`info!`], [`debug!`], and [`trace!`] macros check
> /// this value and discard any message logged at a higher level. The maximum
> /// log level is set by the [`set_max_level`] function.

so probably we don't even need to check for the level in our own log at all?

Reply:
The [`Log` trait docs](https://docs.rs/log/0.4.6/log/trait.Log.html) imply that we do have to check `enabled()` in our `log()` implementation, but it does seem like `enabled()` could use `log::max_level()` rather than `self.level`. 

A bunch of the existing `log` implementations do store their own `level` (which is where I got the approach from) but it seems to work fine without.&hellip;

Reply:
If I'm reading that doc I mentioned in my previous comment, you don't need to check for the level at all -- the macros themselves do it (https://docs.rs/log/0.4.6/src/log/macros.rs.html#36 seems to confirm it).

Reply:
The [docs](https://docs.rs/log/0.4.6/log/trait.Log.html#tymethod.log) for `Log::log` say "Note that `enabled` is not necessarily called before this method. Implementations of log should perform all necessary filtering internally.", which seems to contradict that, so maybe safest to check anyway.